Memory issue(?) with remote_write on ARMv7 with

(Somewhat of a follow-up on #2663)

**What did you do?**

I'm trying to send logs remotely using `remote_write` on ARMv7. 

**What did you expect to see?**

I expected Prometheus to send a payload to the URL specified in `remote_write`.

**What did you see instead? Under which circumstances?**

All. It works fine if I remove the `remote_write` stanza.

**Environment**

A Raspberry Pi 3 Model B running ResinOS.

In case the memory allocation is a concern, here is a dump of that:

```
# free -m
             total       used       free     shared    buffers     cached
Mem:           972        541        431         49         60        264
-/+ buffers/cache:        215        756
Swap:            0          0          0
```

* System information:

Linux 4.4.48 armv7l

* Prometheus version: 

```
prometheus, version 1.6.1 (branch: master, revision: 4666df502c0e239ed4aa1d80abbbfb54f61b23c3)
  build user:       root@ca1b452514b7
  build date:       20170419-13:53:08
  go version:       go1.8.1
```

* Prometheus configuration file:
```
# cat /etc/prometheus.yml
# my global config
global:
  scrape_interval:     30s # Set the scrape interval to every 15 seconds. Default is every 1 minute.
  evaluation_interval: 30s # Evaluate rules every 15 seconds. The default is every 1 minute.
  # scrape_timeout is set to the global default (10s).

  # Attach these labels to any time series or alerts when communicating with
  # external systems (federation, remote storage, Alertmanager).
  external_labels:
      monitor: 'tmp-resin'

# Load rules once and periodically evaluate them according to the global 'evaluation_interval'.
rule_files:
  # - "first.rules"
  # - "second.rules"

# A scrape configuration containing exactly one endpoint to scrape:
# Here it's Prometheus itself.
scrape_configs:
  - job_name: 'prometheus'
    static_configs:
      - targets: ['localhost:9090']
  - job_name: 'tmp'
    static_configs:
      - targets: ['localhost:8000']

remote_write:
    - url: https://requestb.in/abc123
```

* Logs:
```
# prometheus -config.file=/etc/prometheus.yml
INFO[0000] Starting prometheus (version=1.6.1, branch=master, revision=4666df502c0e239ed4aa1d80abbbfb54f61b23c3)  source=main.go:88
INFO[0000] Build context (go=go1.8.1, user=root@ca1b452514b7, date=20170419-13:53:08)  source=main.go:89
INFO[0000] Loading configuration file /etc/prometheus.yml  source=main.go:251
INFO[0000] Loading series map and head chunks...         source=storage.go:421
INFO[0000] 0 series loaded.                              source=storage.go:432
INFO[0000] Starting target manager...                    source=targetmanager.go:61
INFO[0000] Listening on :9090                            source=web.go:259
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x4 pc=0x11580]

goroutine 182 [running]:
sync/atomic.addUint64(0x12448474, 0x1, 0x0, 0x1035a2c, 0x1244844c)
        /usr/local/go/src/sync/atomic/64bit_arm.go:31 +0x4c
github.com/prometheus/prometheus/storage/remote.(*ewmaRate).incr(0x12448474, 0x1, 0x0)
        /go/src/github.com/prometheus/prometheus/storage/remote/ewma.go:65 +0x30
github.com/prometheus/prometheus/storage/remote.(*shards).enqueue(0x123175c0, 0x1250f060, 0x0)
        /go/src/github.com/prometheus/prometheus/storage/remote/queue_manager.go:427 +0x38
github.com/prometheus/prometheus/storage/remote.(*QueueManager).Append(0x12448400, 0x1250ed20, 0x1265e960, 0x0)
        /go/src/github.com/prometheus/prometheus/storage/remote/queue_manager.go:239 +0x190
github.com/prometheus/prometheus/storage/remote.(*Writer).Append(0x1265e960, 0x1250ed20, 0x0, 0x0)
        /go/src/github.com/prometheus/prometheus/storage/remote/write.go:79 +0x8c
github.com/prometheus/prometheus/storage.Fanout.Append(0x124c6c80, 0x2, 0x2, 0x1250ed20, 0x123902f0, 0x125d9338)
        /go/src/github.com/prometheus/prometheus/storage/storage.go:60 +0x58
github.com/prometheus/prometheus/storage.(*Fanout).Append(0x124c71a0, 0x1250ed20, 0x14047d14, 0x123902d3)
        <autogenerated>:3 +0x64
github.com/prometheus/prometheus/retrieval.(*countingAppender).Append(0x126da0a0, 0x1250ed20, 0x14047cfc, 0x125d9378)
        /go/src/github.com/prometheus/prometheus/retrieval/target.go:261 +0x3c
github.com/prometheus/prometheus/retrieval.ruleLabelsAppender.Append(0x1e053b8, 0x126da0a0, 0x1254c140, 0x1250ed20, 0x126da0c0, 0x1265340)
        /go/src/github.com/prometheus/prometheus/retrieval/target.go:201 +0x138
github.com/prometheus/prometheus/retrieval.(*ruleLabelsAppender).Append(0x126da0c0, 0x1250ed20, 0x1e06ad8, 0x126da0c0)
        <autogenerated>:35 +0x64
github.com/prometheus/prometheus/retrieval.(*scrapeLoop).append(0x12328140, 0x124be700, 0xb, 0xc8, 0xe, 0x256ae0f6, 0x1eabaa0)
        /go/src/github.com/prometheus/prometheus/retrieval/scrape.go:519 +0x224
github.com/prometheus/prometheus/retrieval.(*scrapeLoop).run(0x12328140, 0xfc23ac00, 0x6, 0x540be400, 0x2, 0x0)
        /go/src/github.com/prometheus/prometheus/retrieval/scrape.go:429 +0x4e0
created by github.com/prometheus/prometheus/retrieval.(*scrapePool).sync
        /go/src/github.com/prometheus/prometheus/retrieval/scrape.go:258 +0x330
```

Oh man, even the unit test reproduce the issue:

```
$ go test
--- FAIL: TestSampleDelivery (4.83s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x4 pc=0x126b4]

goroutine 28 [running]:
testing.tRunner.func1(0x1076a180)
	/usr/local/go/src/testing/testing.go:622 +0x254
panic(0x28ebd0, 0x425250)
	/usr/local/go/src/runtime/panic.go:489 +0x288
sync/atomic.addUint64(0x10794074, 0x1, 0x0, 0x243150, 0x1079404c)
	/usr/local/go/src/sync/atomic/64bit_arm.go:31 +0x4c
github.com/prometheus/prometheus/storage/remote.(*ewmaRate).incr(0x10794074, 0x1, 0x0)
	/home/pi/go/src/github.com/prometheus/prometheus/storage/remote/ewma.go:65 +0x30
github.com/prometheus/prometheus/storage/remote.(*shards).enqueue(0x107f4150, 0x136fd900, 0x0)
	/home/pi/go/src/github.com/prometheus/prometheus/storage/remote/queue_manager.go:427 +0x38
github.com/prometheus/prometheus/storage/remote.(*QueueManager).Append(0x10794000, 0x107f2720, 0x64, 0x2a05f200)
	/home/pi/go/src/github.com/prometheus/prometheus/storage/remote/queue_manager.go:239 +0x190
github.com/prometheus/prometheus/storage/remote.TestSampleDelivery(0x1076a180)
	/home/pi/go/src/github.com/prometheus/prometheus/storage/remote/queue_manager_test.go:106 +0x3e0
testing.tRunner(0x1076a180, 0x2dcac8)
	/usr/local/go/src/testing/testing.go:657 +0x88
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:697 +0x25c
exit status 2
FAIL	github.com/prometheus/prometheus/storage/remote	4.923s
```

Bring on the ARM CI...

Okay, its embedded struct alignment:

```
package main

import (
        "sync/atomic"
)

type a struct {
        counter int64
}

type b struct {
        b uint32
        a a
}

func main() {
        foo := &b{}
        atomic.AddInt64(&foo.a.counter, 1)
}
```

```
$ go run test.go
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x4 pc=0x11204]

goroutine 1 [running]:
sync/atomic.addUint64(0x1030a0a4, 0x1, 0x0, 0x103000f0, 0x103000f0)
	/usr/local/go/src/sync/atomic/64bit_arm.go:31 +0x4c
main.main()
	/home/pi/test.go:18 +0x50
exit status 2
```

And it seems like multiple embedded structs are tightly-packed:

```
package main

import (
        "sync/atomic"
)

type a struct {
        counter int64
        a uint32
}

type b struct {
        a1 a
        a2 a
        b uint32
}

func main() {
        foo := &b{}
        atomic.AddInt64(&foo.a1.counter, 1)
        atomic.AddInt64(&foo.a2.counter, 1)
}
```

```
$ go run test.go
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x4 pc=0x11204]

goroutine 1 [running]:
sync/atomic.addUint64(0x1036600c, 0x1, 0x0, 0x1, 0x0)
	/usr/local/go/src/sync/atomic/64bit_arm.go:31 +0x4c
main.main()
	/home/pi/test.go:21 +0x60
exit status 2
```

TBH I don't see a want of guaranteeing alignment that isn't super fragile!

--- a/storage/remote/ewma.go
+++ b/storage/remote/ewma.go
@@ -29,8 +29,10 @@ type ewmaRate struct {
 	mutex     sync.Mutex
 }
 
-func newEWMARate(alpha float64, interval time.Duration) ewmaRate {
-	return ewmaRate{
+// newEWMARate always allocates a new ewmaRate, as this guarantees the atomically
+// accessed int64 will be aligned on ARM.  See prometheus#2666.
+func newEWMARate(alpha float64, interval time.Duration) *ewmaRate {
+	return &ewmaRate{
 		alpha:    alpha,
 		interval: interval,
 	}

--- a/storage/remote/queue_manager.go
+++ b/storage/remote/queue_manager.go
@@ -185,7 +185,7 @@ type QueueManager struct {
 	quit        chan struct{}
 	wg          sync.WaitGroup
 
-	samplesIn, samplesOut, samplesOutDuration ewmaRate
+	samplesIn, samplesOut, samplesOutDuration *ewmaRate
 	integralAccumulator                       float64
 }
 

